lib/sudo-pam/src/lib.rs

```diff
@@ -99,8 +90,14 @@ impl<C: Converser> PamContextBuilder<C> {
             if pamh.is_null() {
                 Err(PamError::InvalidState)
             } else {
-                context.pamh = Some(unsafe { &mut *pamh });
-                Ok(context)
+                Ok(PamContext {
+                    data_ptr,
+                    pamh: unsafe { &mut *pamh },
```

Comment:
Would you please explain why this is safe?

Reply:
Firstly, we've passed a pointer to null that was changed by `pam_start` to something else because we check that it is no longer null. Secondly we check there were no errors given by `pam_start`. Those two facts combined should be enough to confirm that a valid pamh pointer is now in `pamh` and can thus safely be converted to a reference. We then convert to a reference only to make sure that there won't be any ownership or lifetime issues going forward (i.e. we can't just leak the ptr and get into all the nasty memory safety issues). The FFI functions that we call do indeed take a ptr, but passing a ref makes no runtime difference, and we might as well have a safe reference instead of a possibly unsafe ptr.